### PR TITLE
Stop emitting non-power-of-two vectors in (non-portable-SIMD) codegen

### LINKED ISSUE #115212: Nearly 50% performance regression when copying/moving small byte arrays between 1.71 and 1.72
### Summary

Nearly 50% performance regression between versions 1.71 and 1.72 when copying small byte arrays with `stable-aarch64-apple-darwin` platform (Macbook Pro M1). Used `cargo bisect-rustc` to track down the regression to fd9bf594366e73efb1a26a023e0b4de8eff82b94 #111999

### Code

I tried this code:

```rust
#![feature(test)]
extern crate test;
use test::Bencher;

#[bench]
fn bench(b: &mut Bencher) {
    b.iter(|| {
        let data = vec![[1u8; 24]; 1024 * 1024];
        let mut new_vec = Vec::with_capacity(1024 * 1024);
        for entry in data {
            new_vec.push(entry);
        }
    });
}
```

I expected to see this happen: to run about 6,200,000 ns/iter like version 1.71

```
$ cargo +nightly-2023-06-06-aarch64-apple-darwin bench
   Compiling benchmarks v0.1.0 (/Users/tuomas/rust/benchmarks)
    Finished bench [optimized] target(s) in 0.18s
     Running unittests src/main.rs (target/release/deps/benchmarks-0c0dfc3834e16b96)

running 1 test
test bench ... bench:   6,249,724 ns/iter (+/- 1,609,874)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out; finished in 5.70s
```

Instead, this happened: 9,159,412 ns/iter

```
$ cargo +nightly-2023-06-07-aarch64-apple-darwin bench
    Finished bench [optimized] target(s) in 0.00s
     Running unittests src/main.rs (target/release/deps/benchmarks-0c0dfc3834e16b96)

running 1 test
test bench ... bench:   9,159,412 ns/iter (+/- 469,568)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out; finished in 2.78s
```

### Version it worked on

It most recently worked on: 1.71
More specifically nightly-2023-06-06-aarch64-apple-darwin

### Version with regression

1.72
nightly-2023-06-07-aarch64-apple-darwin

`rustc --version --verbose`:
```
rustc 1.72.0 (5680fa18f 2023-08-23)
binary: rustc
commit-hash: 5680fa18feaa87f3ff04063800aec256c3d4b4be
commit-date: 2023-08-23
host: aarch64-apple-darwin
release: 1.72.0
LLVM version: 16.0.5
```

### cargo bisect-rustc

Searched nightlies: from nightly-2023-06-06 to nightly-2023-06-07
regressed nightly: nightly-2023-06-07
searched commit range: https://github.com/rust-lang/rust/compare/e6d4725c76f3b526c74454bc51afdf6daf133506...b2b34bd83192c3d16c88655158f7d8d612513e88
regressed commit: https://github.com/rust-lang/rust/commit/fd9bf594366e73efb1a26a023e0b4de8eff82b94

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>

Host triple: x86_64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --start=2023-06-06 --end=2023-06-07 -- run --release
```
</details>

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

### PR DESCRIPTION
**Pull request by scottmcm**
Fixes #115212 

It's unclear what makes this not work sometimes, since it often *does* work, so for now just disable the unusual cases.  A future PR can consider doing something smarter, but this is an easy and safe tweak that we can do to resolve the regressions for now.

### CODE REVIEW

File: tests/codegen/mem-replace-simple-type.rs [hunk 1]
```diff
@@ -33,12 +33,21 @@ pub fn replace_ref_str<'a>(r: &mut &'a str, v: &'a str) -> &'a str {
...
+}
+
+#[no_mangle]
+// CHECK-LABEL: @replace_short_array_4(
+pub fn replace_short_array_4(r: &mut [u32; 4], v: [u32; 4]) -> [u32; 4] {
+    // CHECK-NOT: alloca
+    // CHECK: %[[R:.+]] = load <4 x i32>, ptr %r, align 4
+    // CHECK: store <4 x i32> %[[R]], ptr %result
+    // CHECK: %[[V:.+]] = load <4 x i32>, ptr %v, align 4
+    // CHECK: store <4 x i32> %[[V]], ptr %r
```
Comment (workingjubilee) [inline tests/codegen/mem-replace-simple-type.rs:51]: RIP.

**Review by compiler-errors — APPROVED** [review]
r=me unless you want a specific review

**Comment by Nilstrieb** [comment]
note that this is also what killed the MIPS backend so it seems reasonable to believe that LLVM backends are not good at this in general

**Comment by scottmcm** [comment]
## Beta nomination

This is to address the perf hit seen in https://github.com/rust-lang/rust/issues/115212

It should be very safe, since it's only one line of product code, and all it does is disable #111999 in more cases, leaving those cases to use the same codepath as in 1.71.

(It would also be possible to disable 111999 *entirely*, but I've also seen cases like <https://rust.godbolt.org/z/7s9ob85ja> where the vector codegen helps avoid byte-by-byte `mov`s, so this targets the known-bad case.  As Nils mentions in https://github.com/rust-lang/rust/pull/115236#issuecomment-1694205112, non-power-of-two vectors also seem to be less well exercised in backends, so this avoids those problems by only using power-of-two vectors.)

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-08-31/near/388333642)

Also in favor of a stable backport in case a 1.72.1 materializes ([zulip comment](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-08-31/near/388337969)).

@rustbot label +beta-accepted +stable-accepted

**Pull request merged**